src/write.jl

```diff
@@ -90,35 +101,33 @@ end
 # tp_tensor_elemtype(::Type{Float32}) = Integer(var"TensorProto.DataType".FLOAT)
 # tp_tensor_elemtype(::Type{Float64}) = Integer(var"TensorProto.DataType".DOUBLE)
 
-TensorProto(x::Number, name ="") = TensorProto([x], name)
-
-TensorProto(t::AbstractArray{Float64,N}, name ="") where N = TensorProto(
-    dims=collect(reverse(size(t))),
-    data_type=var"TensorProto.DataType".DOUBLE |> Integer,
-    double_data = reshape(t, :),
-    name=name)
-
-TensorProto(t::AbstractArray{Float32,N}, name ="") where N = TensorProto(
-    dims=collect(reverse(size(t))),
-    data_type=var"TensorProto.DataType".FLOAT |> Integer,
-    float_data = reshape(t, :),
-    name=name)
-
-TensorProto(t::AbstractArray{Float16,N}, name ="") where N = TensorProto(t, var"TensorProto.DataType".FLOAT16 |> Integer, name)
-
-TensorProto(t::AbstractArray{Int64,N}, name ="") where N = TensorProto(
-    dims=collect(reverse(size(t))),
-    data_type=var"TensorProto.DataType".INT64 |> Integer,
-    int64_data = reshape(t, :),
-    name=name)
+# TensorProto interface
+# TensorProto(t::AbstractArray{T,N}, name ="")
+# should follow onnx.proto3 (v1.13.1) 
+for (T, field) in [(Float32, :float_data)
+                   (UInt8, :raw_data)
+                   (Int32, :int32_data)
+                   (Int64, :int64_data)
+                   (String, :string_data)
+                   (Bool , :int32_data)
+                   (Float64, :double_data)
+                   (UInt64, :uint64_data)]
+    @eval TensorProto(t::AbstractArray{$T,N}, name ="") where N = TensorProto(
+          dims=collect(reverse(size(t))),
+          data_type=elem_type_code($T),
+          $(field) = reshape(t, :),
+          name=name)
+end
 
-TensorProto(t::AbstractArray{Int32,N}, name ="") where N = TensorProto(
-    dims=collect(reverse(size(t))),
-    data_type=var"TensorProto.DataType".INT32 |> Integer,
-    int32_data = reshape(t, :),
-    name=name)
+for T in [Int8, UInt16, Int16, Float16, UInt32]#, Bool,
+    @eval TensorProto(t::AbstractArray{$T,N}, name ="") where N = TensorProto(
+          dims=collect(reverse(size(t))),
+          data_type=elem_type_code($T),
+          raw_data = reinterpret(UInt8, reshape(t, :)),
+          name=name)
+end
 
-TensorProto(t::AbstractArray{Int8,N}, name ="") where N = TensorProto(t, var"TensorProto.DataType".INT8 |> Integer, name)
+TensorProto(x::Number, name ="") = TensorProto([x], name)
```

Comment:
I'm curious what's the motivational case here? Is it just for convenience or you encountered a case where a number in Julia mapped to a single-item array in ONNX?